src/charm.py

```diff
@@ -466,6 +468,54 @@ def _on_create_backup_action(self, event: ActionEvent) -> None:
         logger.info("Backup uploaded to S3 bucket %s", s3_parameters["bucket"])
         event.set_results({"backup-id": backup_key})
 
+    def _on_list_backups_action(self, event: ActionEvent) -> None:
+        """Handle the list-backups action.
+
+        Lists all backups stored in S3 bucket.
+
+        Args:
+            event: ActionEvent
+        """
+        if not self.unit.is_leader():
+            event.fail(message="Only leader unit can perform backup operations.")
+            logger.error("Failed to run list-backups action - Only leader unit can perform backup operations.")
+            return
+
+        if not self._is_relation_created(S3_RELATION_NAME):
+            event.fail(message="Failed to list backups. S3 relation not created.")
+            logger.error("Failed to run list-backups action - S3 relation not created.")
+            return
+
+        if missing_parameters:= self._get_missing_s3_parameters():
+            event.fail(message=f"Failed to list backups. S3 parameters missing: {missing_parameters}")
+            logger.error("Failed to run list-backups action - S3 parameters missing.")
+            return
+
+        s3_parameters = self._get_s3_parameters()
+
+        try:
+            s3 = S3(
+                access_key=s3_parameters["access-key"],
+                secret_key=s3_parameters["secret-key"],
+                endpoint=s3_parameters["endpoint"],
+                region=s3_parameters.get("region"),
+            )
+        except S3Error:
+            event.fail(message="Failed to create S3 session.")
+            logger.error("Failed to run list-backups action - Failed to create S3 session.")
```

Comment:
Should we log the exception here? If this happens, we would not be able to debug without further information from the exception.

Reply:
Done